### ISSUE TITLE
Reduce false positives on subsequent additions to heterogeneous dictionaries

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#2971]
These are cases which would ideally be typed as `TypedDict`, but aren't. For example:

```py
def make_person(photo: bytes | None):
    person = {"name": "Pat", age: 29}
    if photo is not None:
        person["photo"] = photo
```

We (and pyrefly, and pyright in strict mode) error on the last line here because we already inferred `dict[str, str | int]`, so we can't add a `bytes` value.

Mypy prefers common-base joins over union joins, so it infers `dict[str, object]`, which avoids the error adding a `bytes` value. This means the value type is less precise, which theoretically means potentially more errors using values from the dict later. But in practice with this heterogeneous pattern, either `object` or the union will cause similar problems when using values from the dict -- in either case you'd probably have to cast or narrow.

Pyright (in non-strict mode) has a special case where it falls back to `Unknown` when it sees heterogenous value types, so it infers this as `dict[str, Unknown]`.

I think we could consider either the mypy or pyright approaches here. I kind of like the mypy approach, because it doesn't introduce a gradual type. The idea would be that in type promotion, we simplify heterogeneous unions to their common nominal base type (which would be `object` if there isn't a more precise one).

To do this, we might have to start tracking unions which originate in a user annotation vs those which don't (similar to how we already do the same for literals), so that we only simplify synthesized unions, not user-annotated ones. This might be generally a good thing to track, though, since it would also allow us to generally do less simplification of user-annotated union types.

**Comment by ibraheemdev** [comment]
I'm a little confused why this is a separate issue than https://github.com/astral-sh/ty/issues/1473. Wouldn't we just infer `person` as `dict[str, str | int | bytes | None]` with full-scope bidirectional inference? Or are you suggesting a simpler temporary fix before that is implemented?

**Comment by carljm** [comment]
That's a fair question! I could have just made this a comment on #1473. I think I had them as more distinct in my mind partly because #1473 is framed entirely around empty containers ("not constrained at construction"). I did mark this as a child issue of #1473.

It's not clear yet how far we'll be able to go with the approach in #1473 without causing too much performance or cycle-related regression. Will we be able to apply it to all container literals, not just empty ones?

It's also not clear to me if applying #1473 to all container literals necessarily has preferable semantics. It could cause actual mistakes (wrong type added to container later) to widen the type instead of erroring, leading to more confusing diagnostics.

I wanted to capture observations about how other type checkers handle cases like this in the ecosystem, to give us a fuller picture of options, in case #1473 doesn't fully solve the problem or we need to do something else before it's ready. But I think it makes sense to explore what we can do with #1473.
